src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs

```diff
@@ -58,39 +48,33 @@ public override void Accept(Chunk chunk)
             var parentChunk = chunk as ParentChunk;
             var tagHelperChunk = chunk as TagHelperChunk;
 
-            if (parentChunk != null && !(parentChunk is TagHelperChunk))
+            if (parentChunk != null)
             {
                 Accept(parentChunk.Children);
             }
 
-            else if (tagHelperChunk != null) 
+            if (tagHelperChunk != null) 
             {
-                var viewComponentDescriptors = tagHelperChunk.Descriptors.Where(
-                    descriptor => ViewComponentTagHelperDescriptorConventions.IsViewComponentDescriptor(descriptor));
-                foreach (var descriptor in viewComponentDescriptors)
-                {
-                    base.Accept(chunk);
-                    return;
-                }
+                base.Accept(chunk);
             }
         }
 
-        /// <summary>
-        /// We generate classes for each view component <see cref="TagHelperDescriptor"/> in the chunk.
-        /// </summary>
-        /// <param name="chunk">The <see cref="TagHelperChunk"/> to visit.</param>
         protected override void Visit(TagHelperChunk chunk)
         {
-            var viewComponentDescriptors = chunk.Descriptors.Where(
-                descriptor => ViewComponentTagHelperDescriptorConventions.IsViewComponentDescriptor(descriptor));
+            var viewComponentDescriptors = chunk.Descriptors.
+                Where(ViewComponentTagHelperDescriptorConventions.IsViewComponentDescriptor);
 
             foreach (var descriptor in viewComponentDescriptors)
             {
-                var descriptorName =
+                var shortName =
                     descriptor.PropertyBag[ViewComponentTagHelperDescriptorConventions.ViewComponentNameKey];
-                if (!_writtenViewComponents.Contains(descriptorName))
+                // TODO: type name generation as a method. 
```

Comment:
We check the type name as `__Generate__{ShortName}ViewComponentTagHelper` across different pull requests. After everything is merged, we'll make this into one method.